### ISSUE TITLE
Sample a lots of invalid actions while training.

### ISSUE THREAD

**Issue by changpowei** [vwxyzjn/gym-microrts#8]
Hi @vwxyzjn , sorry for bothering you again. After study this case entire week, I have a couple of questions ask u. 

1. 
I found the output of `env.action_mask` will get all 0 sometimes, despite there have our units or opponent units on the map. Ideally, I think the location of units should be 1 .

```
invalid_action_masks = torch.Tensor(np.array(env.action_mask)).unsqueeze(0)
```

2.
 In my opinion, Categorical(logits=logits) should only consider 1 while converting into probability. However, it take 0 into the process. 
```
multi_categoricals = [Categorical(logits=logits) for logits in split_invalid_action_masks]
```
Please refer to the figure below.

The map, 
![擷取選取區域_002](https://user-images.githubusercontent.com/22905460/91009536-58fd8c00-e613-11ea-8587-fb2d1527d4fe.png)

Following figure shows that the `multi_categoricals[0].probs` remains have probability 0.0095, despite there is no unit on that location.
![2020-08-24 14-16-02 的螢幕擷圖](https://user-images.githubusercontent.com/22905460/91010252-c0680b80-e614-11ea-9f55-55a4ba5a0d36.png)

I think it should only consider the location where have units, and the probability mask should like this. Or do I misunderstand something?
![2020-08-24 14-57-46 的螢幕擷圖](https://user-images.githubusercontent.com/22905460/91013470-5488a180-e61a-11ea-9f50-1bb7dc110688.png)

But here has a very interesting part, I just split the code
```
 multi_categoricals = [Categorical(logits=logits) for logits in split_invalid_action_masks]
```
into
```
multi_categoricals = []
for logits in split_invalid_action_masks:
            multi_categoricals.append(Categorical(logits))
```
then the result will be the same as I thought.

3. 
After the above steps, convert `multi_categoricals` to the physical action by sampling. I have some questions here. Because every actions or directions have same probability to be sampled, it may appear lots of invalid actions. For example, it may indicate unit Ranged to harvest, or indicate the location where doesn't exist any unit to do some action.  Is this appropriate to indicate the invalid action while training?
```
action = torch.stack([categorical.sample() for categorical in multi_categoricals]).flatten().tolist()
```
I just filter out invalid actions while training. Will this improve training efficiency?

4.
I found when two teams draw, the game continues until reaching max_episode_steps set previously.
```
next_obs, reward, done, _ = env.step(action)
```

5.
Finally, it pops up following error when training the model. But it does not seem to affect training, it can keep training. Can you please explain the reason for this error?
![擷取選取區域_001](https://user-images.githubusercontent.com/22905460/91009134-7aaa4380-e612-11ea-9c8e-a172b37a8da0.png)

very appreciate for ur reply!!!!

**Comment by vwxyzjn** [comment]
Hi @changpowei , I want to first thank you for taking the time to report this issue. I will take a look today. Second, I have just created a discord server in case you wanted to join https://discord.gg/5URpuE, where you can directly ask smaller questions.

**Comment by vwxyzjn** [comment]
Regarding issue 2. The code that calculates the action mask is here:

https://github.com/vwxyzjn/gym-microrts/blob/7ac6f8ad0c6f4bd51836228c9951fa68756710c4/gym_microrts/envs/global_agent_env.py#L63-L69

Action mask is composed of unit location mask, action type and parameters mask, and target unit location mask. So far we have not implemented the action type and parameters mask.
So given a location, essentially the unit location mask creates a mask 1) if there is not a unit, 2) the unit is busy, 3) the unit is not owned by the player 1. However, it is possible to encounter a situation where the unit location mask is all zero even though player 1 has units that exist because all of the units might be busy, which I suspect is the reason for issue 2, where the probability is uniform for unit selection.

**Comment by vwxyzjn** [comment]
Regarding issues 3-4, they seem to be issues related to the implementation of invalid action masking. I suggest looking at our paper https://arxiv.org/abs/2006.14171 and see if that helps.

**Comment by vwxyzjn** [comment]
Lastly issue 5 does not seem to cause too many troubles so I kind of just left them there...

**Comment by vwxyzjn** [comment]
Closing the issue for now but feel free to open another issue should you have more questions.
